File: src/pkjs/quick.js
/* Saved shortcuts to somewhere you go often.
 *
 * ES5 ONLY; see the note at the top of appmessage.js.
 *
 * Browsing to a deep action costs a lot of button presses -- on the homelab's
 * root, the last action is fifteen rows down. A quick item remembers a
 * destination so it can be reached from the opening screen instead.
 *
 * What it remembers is deliberately not the whole answer:
 *
 *  - **An action is stored by name, with the address of the document that
 *    offered it** -- never the action's own href. Actions come and go as
 *    server state changes, and their hrefs are the server's business. Running
 *    one means re-fetching the holder and looking the name up again, exactly
 *    as if you had walked there; if it is no longer offered, that is a real
 *    answer and it is reported.
 *  - **A backend is stored by base URL, not by position.** Reordering the
 *    backend list must not silently re-point a saved shortcut at a different
 *    server.
 *
 * A shortcut is a shortcut through the *navigation*, not through the
 * deciding: an action reached this way still goes through the same
 * confirmation as one reached by hand.
 */

'use strict';

var settings = require('./settings');

var STORAGE_KEY = 'restforge.quick';

/* More than this and the opening screen stops being quicker than browsing. */
var MAX_QUICK = 12;

var KIND_ACTION = 'action';
var KIND_DOCUMENT = 'document';

function trim(value) {
  if (value === null || value === undefined) {
    return '';
  }
  return String(value).replace(/^\s+|\s+$/g, '').slice(0, 256);
}

/* normalise coerces one stored item into the shape above.  Anything that has
 * drifted degrades to something usable rather than throwing during startup --
 * the same reasoning as settings.js. */
function normalise(raw) {
  var item = raw && typeof raw === 'object' ? raw : {};
  return {
    label: trim(item.label),
    backendName: trim(item.backendName),
    baseUrl: trim(item.baseUrl),
    kind: item.kind === KIND_ACTION ? KIND_ACTION : KIND_DOCUMENT,
    holder: trim(item.holder),
    name: trim(item.name),
    href: trim(item.href)
  };
}

/* usable rejects items that could not be acted on.  An action needs somewhere
 * to look itself up; a document needs an address. */
function usable(item) {
  if (!item.label || !item.baseUrl) {
    return false;
  }
  return item.kind === KIND_ACTION ? !!(item.holder && item.name)
                                   : !!item.href;
}

function load() {
  var raw;
  try {
    raw = localStorage.getItem(STORAGE_KEY);
  } catch (error) {
    console.log('quick: localStorage unavailable: ' + error);
    return [];
  }
  if (!raw) {
    return [];
  }
  var parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    console.log('quick: stored shortcuts are not JSON, ignoring: ' + error);
    return [];
  }
  if (Object.prototype.toString.call(parsed) !== '[object Array]') {
    return [];
  }
  var out = [];
  for (var i = 0; i < parsed.length && out.length < MAX_QUICK; i++) {
    var item = normalise(parsed[i]);
    if (usable(item)) {
      out.push(item);
    }
  }
  return out;
}

function save(list) {
  var clean = [];
  for (var i = 0; i < list.length && clean.length < MAX_QUICK; i++) {
    var item = normalise(list[i]);
    if (usable(item)) {
      clean.push(item);
    }
  }
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(clean));
  } catch (error) {
    console.log('quick: could not save: ' + error);
    return load();
  }
  return clean;
}

function count() {
  return load().length;
}

function get(index) {
  var list = load();
  return (index < 0 || index >= list.length) ? null : list[index];
}

/* add appends a shortcut, replacing any identical one rather than
 * accumulating duplicates -- saving the same row twice is a natural thing to
 * do and should be idempotent. */
function add(item) {
  var wanted = normalise(item);
  if (!usable(wanted)) {
    console.log('quick: refusing to save an incomplete shortcut');
    return null;
  }
  var list = load();
  for (var i = 0; i < list.length; i++) {
    if (same(list[i], wanted)) {
      list[i] = wanted;
      save(list);
      return wanted;
    }
  }
  if (list.length >= MAX_QUICK) {
    console.log('quick: already holding ' + MAX_QUICK + ' shortcuts');
    return null;
  }
  list.push(wanted);
  save(list);
  return wanted;
}

function same(a, b) {
  return a.baseUrl === b.baseUrl && a.kind === b.kind &&
         a.holder === b.holder && a.name === b.name && a.href === b.href;
}

function remove(index) {
  var list = load();
  if (index < 0 || index >= list.length) {
    return false;
  }
  list.splice(index, 1);
  save(list);
  return true;
}

/* backendFor resolves a shortcut against the configured backends, by base URL
 * so that renaming or reordering them does not re-point it somewhere else.
 * Returns null when the backend it referred to is gone. */
function backendFor(item) {
  var backends = settings.load();
  for (var i = 0; i < backends.length; i++) {
    if (backends[i].baseUrl === item.baseUrl) {
      return backends[i];
    }
  }
  return null;
}

/* rows renders the Quick section for the opening screen.  A shortcut whose
 * backend has been deleted is still shown, marked -- silently dropping
 * something the user saved would leave them wondering where it went. */
function rows() {
  var list = load();
  var out = [];
  for (var i = 0; i < list.length; i++) {
    var backend = backendFor(list[i]);
    out.push({
      label: list[i].label,
      sublabel: backend ? backend.name : 'backend removed',
      kind: 'q'
    });
  }
  return out;
}

module.exports = {
  KIND_ACTION: KIND_ACTION,
  KIND_DOCUMENT: KIND_DOCUMENT,
  MAX_QUICK: MAX_QUICK,
  load: load,
  save: save,
  count: count,
  get: get,
  add: add,
  remove: remove,
  backendFor: backendFor,
  rows: rows
};

File: cli/internal/quick/quick.go
package quick

import "strings"

// MaxQuick caps how many shortcuts may be saved. More than this and the
// opening screen stops being quicker than browsing. Matches
// QuickService.maxQuick (quick_service.dart) and MAX_QUICK (quick.js).
const MaxQuick = 12

// MaxFieldLength caps every free-text field, for the same reason
// backend.MaxFieldLength does -- generous, because the limit that actually
// matters is screen space. Matches QuickService.maxFieldLength.
const MaxFieldLength = 256

// QuickKind is what a shortcut points at. Mirrors KIND_ACTION/KIND_DOCUMENT
// in quick.js, as a closed Go type rather than a pair of string constants.
type QuickKind int

const (
	// KindAction is a shortcut to an action, looked up by Holder and Name
	// at run time -- see the package comment for why it never stores the
	// action's own href.
	KindAction QuickKind = iota
	// KindDocument is a shortcut to a document, fetched by Href.
	KindDocument
)

// String names a QuickKind the way it is stored on disk ("action" or
// "document"). Any value other than KindAction reads as "document",
// matching the ternary quick.js's normalise and this package's own
// Normalise use to coerce a drifted or hand-edited Kind.
func (k QuickKind) String() string {
	if k == KindAction {
		return "action"
	}
	return "document"
}

// QuickItem is one saved shortcut. A plain value with no behaviour beyond
// reading and persisting itself, the Go counterpart to quick_service.dart's
// QuickItem class.
//
// Holder and Name are what an action shortcut is looked up by; Href is
// what a document shortcut is fetched from -- see the package comment for
// why an action never stores an href of its own. BackendName is the
// backend's display name at the moment the shortcut was saved: kept for
// schema parity with quick.js (which stores it the same way) even though
// nothing in this package reads it back -- BackendFor/BackendsFor resolve
// the current name by BaseURL instead, precisely so a rename is reflected
// rather than frozen at save time.
type QuickItem struct {
	Label       string
	BackendName string
	BaseURL     string
	Kind        QuickKind
	Holder      string
	Name        string
	Href        string
}

// Normalise coerces item into the shape Load/Save/Add expect: every
// free-text field trimmed of surrounding whitespace and capped at
// MaxFieldLength, and Kind coerced to a known value (see QuickKind.String).
// It does not reject anything -- usable does that. Input that has drifted
// (an older layout, a hand-edited config file) degrades to something
// usable instead of taking a caller down. Mirrors quick_service.dart's
// normalise/quick.js's normalise.
func Normalise(item QuickItem) QuickItem {
	kind := KindDocument
	if item.Kind == KindAction {
		kind = KindAction
	}
	return QuickItem{
		Label:       trim(item.Label),
		BackendName: trim(item.BackendName),
		BaseURL:     trim(item.BaseURL),
		Kind:        kind,
		Holder:      trim(item.Holder),
		Name:        trim(item.Name),
		Href:        trim(item.Href),
	}
}

// usable rejects a shortcut that could not be acted on. An action needs
// somewhere to look itself up next time; a document needs an address.
// Mirrors usable() in quick_service.dart/quick.js.
func usable(item QuickItem) bool {
	if item.Label == "" || item.BaseURL == "" {
		return false
	}
	if item.Kind == KindAction {
		return item.Holder != "" && item.Name != ""
	}
	return item.Href != ""
}

// same reports whether a and b point at the same target -- used by Add to
// decide whether saving is an update rather than a new entry. Mirrors
// same() in quick_service.dart/quick.js.
func same(a, b QuickItem) bool {
	return a.BaseURL == b.BaseURL &&
		a.Kind == b.Kind &&
		a.Holder == b.Holder &&
		a.Name == b.Name &&
		a.Href == b.Href
}

// trim mirrors backend.clean/QuickService._trim: surrounding whitespace
// stripped, capped at MaxFieldLength runes (not bytes, so truncation can
// never split a multi-byte UTF-8 character and leave invalid UTF-8
// behind).
func trim(s string) string {
	s = strings.TrimSpace(s)
	r := []rune(s)
	if len(r) > MaxFieldLength {
		r = r[:MaxFieldLength]
	}
	return string(r)
}
